packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -1128,9 +1135,61 @@ class _CupertinoActionSheetActionState extends State<CupertinoActionSheetAction>
     widget.onPressed();
   }
 
+  // Calculates the font size for action sheet buttons.
+  //
+  // The `contextBodySize` is the body font size specified by context. The
+  // return value is the button font size, including the effect of context font
+  // scale factor. Divide by context font scale factor before using in a `Text`.
+  static double _buttonFontSize(double contextBodySize) {
+    // It is observed that the native action sheet buttons use font sizes that
+    // deviate from standard HIG specifications in a non-linear way. The following
+    // table shows the regular body font size vs the button font size:
+    //
+    //  Text scale  | xs |  s |  m |  l | xl | xxl | xxxl | ax1 | ax2 | ax3 | ax4 | ax5
+    //  Body font   | 14 | 15 | 16 | 17 | 19 |  21 |  23  |  28 |  33 |  40 |  47 |  53
+    //  Button font | 21 | 21 | 21 | 21 | 23 |  24 |  24  |  28 |  33 |  40 |  47 |  53
+
+    // For very small or very large text, simple rules are applied.
+    //
+    // The upper bound is chosen to be closer to 24, instead of 28, because a
+    // shorter range between the bounds allows a closer fit within them (as seen
+    // later).
+    const double lowerBound = 17.0;
+    const double upperBound = 24.5;
+
+    const double minSize = 21.0;
+
+    return switch (contextBodySize) {
+      < lowerBound => minSize,
+
+      > upperBound => contextBodySize,
+
+      // For mid-sized text, the following formula is used.
+      // This formula is derived through quadratic fitting of the following
+      // table, which is part of the overall table above.
+      //
+      //  x | 17 | 19 |  21 |  23  |  24.5
+      //  y | 21 | 23 |  24 |  24  |  24.5
```

Comment:
This last column doesn't look like it can be part of the same quadratic as the other columns. It has y increasing with x again, after it's already stopped increasing earlier.

Reply:
A quadratic that fits the remaining columns is going to have the property that $y$ is actually greatest at $x = 22$ — it'll be slightly above 24 — and then decreases as $x$ gets bigger from there. That feels unnatural.

My suggestion would be to use a piecewise-linear fit the whole way: so one line segment from 17 to 19, another from 19 to 21, another from 21 to I guess 24, and then it's just `contextBodySize` from there.

Reply:
(Related Discord thread: https://discord.com/channels/608014603317936148/608018585025118217/1260034530870300778 )